searx/version.py

```diff
@@ -1,26 +1,150 @@
+
+# some code is from https://github.com/pypa/setuptools_scm/blob/b0b46fd723f6649a7ba4895b005bc89db1529e22/src/setuptools_scm/utils.py  # pylint: disable=line-too-long
+# ( MIT License )
+
+import os
+import shlex
+import subprocess
+
+
+class ShellException(Exception):
+
+    def __init__(self, command, exit_code, stdout, stderr):
+        super().__init__(self, stderr)
+        self.command = command
+        self.exit_code = exit_code
+        self.stdout = stdout
+        self.stderr = stderr
+
+
+def no_git_env(env):
+    # adapted from pre-commit
+    # Too many bugs dealing with environment variables and GIT:
+    # https://github.com/pre-commit/pre-commit/issues/300
+    # In git 2.6.3 (maybe others), git exports GIT_WORK_TREE while running
+    # pre-commit hooks
+    # In git 1.9.1 (maybe others), git exports GIT_DIR and GIT_INDEX_FILE
+    # while running pre-commit hooks in submodules.
+    # GIT_DIR: Causes git clone to clone wrong thing
+    # GIT_INDEX_FILE: Causes 'error invalid object ...' during commit
+    return {
+        k: v
+        for k, v in env.items()
+        if not k.startswith("GIT_")
+        or k in ("GIT_EXEC_PATH", "GIT_SSH", "GIT_SSH_COMMAND")
+    }
+
+
+def _popen_pipes(cmd, cwd):
+    # pylint: disable=consider-using-with
+    return subprocess.Popen(
+        cmd,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        cwd=str(cwd),
+        env=dict(
+                no_git_env(os.environ),
+                # os.environ,
+                # try to disable i18n
+                LC_ALL="C",
+                LANGUAGE="",
+        ),
+    )
+
+
+def ensure_stripped_str(str_or_bytes):
+    if isinstance(str_or_bytes, str):
+        return str_or_bytes.strip()
+    return str_or_bytes.decode("utf-8", "surrogateescape").strip()
+
+
+def do_ex(cmd, cwd="."):
+    if os.name == "posix" and not isinstance(cmd, (list, tuple)):
+        cmd = shlex.split(cmd)
+
+    p = _popen_pipes(cmd, cwd)
+    out, err = p.communicate()
+    return ensure_stripped_str(out), ensure_stripped_str(err), p.returncode
+
+
+def do_ex_safe(cmd, cwd="."):
+    out, err, exit_code = do_ex(cmd, cwd)
+    if exit_code != 0:
+        raise ShellException(cmd, exit_code, out, err)
+    return out
+
+
+def get_public_git_url(git_url):
+    if git_url.startswith('git@'):
+        git_url = git_url.replace(':', '/', 2).replace('git@', 'https://', 1)
+    if git_url.endswith('.git'):
+        return git_url.replace('.git', '', 1)
+    return git_url
+
+
+def get_git_url_and_branch():
+    try:
+        ref = do_ex_safe('git rev-parse --abbrev-ref @{upstream}')
+    except ShellException:
+        ref = do_ex_safe('git rev-parse --abbrev-ref master@{upstream}')
+    origin, git_branch = ref.split('/', 1)
+    git_url = do_ex_safe(f'git remote get-url {origin}')
+    return get_public_git_url(git_url), git_branch
+
+
+def get_git_version():
+    git_version = do_ex_safe(r'git describe --match "v[0-9]*\.[0-9]*\.[0-9]*" HEAD')
```

Comment:
I think we should not use strings to start a command line tool (needs be executed in a shell / or like you do, using `shlex.split(cmd)` ).  For the sake of simplicity I implemented the [fspath.callEXE](https://return42.github.io/fspath/fspath-api/fspath.fspath.html#fspath.fspath.callEXE). 
```python
>>> import fspath
>>> fspath.callEXE('git', 'describe')
('v1.0.0-590-g0686e274\n', '', 0)
```

If you do not want to use fspath (which is ATM only in requirements-dev.txt) you can use:

```python
>>> import subprocess
>>> proc = subprocess.run(['git', 'describe'], stdout=subprocess.PIPE)
>>> proc.stdout.decode()
'v1.0.0-590-g0686e274\n'

```

Reply:
> I think we should not use strings to start a command line tool (needs be executed in a shell / or like you do, using shlex.split(cmd) ).

I don't know understand why `shlex.split(cmd)` is a problem ( it is a copy / paste from setup_scm ).
For me, 
`'git describe --match "v[0-9]*\.[0-9]*\.[0-9]*" HEAD'` 
is more readable than
`['git', 'describe', '--match', 'v[0-9]*\\.[0-9]*\\.[0-9]*', 'HEAD']`

Reply:
Not really a problem but I would prefer a notation which do not need any kind of *syntax parser".  If you prefer to use one string instead of a argument list you can set `shell=True` in the `subprocess.*` functions (no need for shlex).  Nevertheless ..
[Providing a sequence of arguments is generally preferred, as it allows the module to take care of any required escaping and quoting of arguments (e.g. to permit spaces in file names)](https://docs.python.org/3/library/subprocess.html#frequently-used-arguments)

Reply:
* `shell=True` is a different thing: it calls a shell to actually call the command.
* in `version.py`, all commands are hard coded except one.

Reply:
> `shell=True` is a different thing: it calls a shell to actually call the command.

Maybe I am used to it differently: for me, when I see a command line in one string I suppose this line will be executed through the shell (including wildcards and other shell builtins) and when I see a list of arguments I suppose that a "program" _is executed_ directly in a subprocess. 

> in `version.py`, all commands are hard coded except one.

I do not understand what you mean by this sentence  .. what means "hard coded" here and "except one" .. all commands in version.py are _hard coded_ (in the meaning of: a constant string is used)

Reply:
Is it acceptable to have to string in this particular use case?

> all commands in version.py are hard coded 

yes except:
https://github.com/searxng/searxng/blob/e5d1cd3c9deccfd2099e48baed855dcc00113ed0/searx/version.py#L79

The shell is not involved only `git`, and `origin` comes from a previous git output.